contracts/Enforcer.sol

```diff
@@ -1,134 +1,184 @@
-        bondAmount = _bondAmount;
-        maxExecutionDepth = _maxExecutionDepth;
+        return taskHash;
     }
 
-    // register a new execution
     function register(
-        address codeContractAddress,
-        bytes memory _callData,
-        bytes32 endHash,
-        uint256 executionDepth,
-        // optional
-        bytes32 customEnvironmentHash
-    )
-        public payable
+        bytes32 _taskHash,
+        bytes32 _solverPathRoot,
+        bytes32[] memory _resultProof,
+        bytes memory _result
+    ) public payable
     {
+        uint256 executionDepth = _resultProof.length;
+        bytes32 executionId = keccak256(abi.encodePacked(_taskHash, _solverPathRoot));
+
+        ExecutionResult storage executionResult = executions[executionId];
+        Task storage task = tasks[_taskHash];
+
+        require(task.startBlock != 0, "Execution request does not exist");
+        //require(task.startBlock + taskPeriod > block.number, "Task period is over");
+        require((block.number + (challengePeriod * 2)) < (task.startBlock + taskPeriod), "Too late for registration");
+        require(executionResult.startBlock == 0, "Execution already registered");
         require(msg.value == bondAmount, "Bond is required");
         require(executionDepth <= maxExecutionDepth, "Execution too long");
 
-        bytes32 executionId = keccak256(abi.encodePacked(codeContractAddress, _callData));
+        executionResult.startBlock = block.number;
+        executionResult.taskHash = _taskHash;
+        executionResult.executionDepth = executionDepth;
+        executionResult.solverPathRoot = _solverPathRoot;
+        executionResult.resultHash = keccak256(abi.encodePacked(_result));
+        executionResult.solver = msg.sender;
+
+        task.executions.push(executionId);
+
+        bonds[msg.sender] += msg.value;
 
-        require(executions[executionId].startBlock == 0, "Execution already registered");
-        executions[executionId] = Execution(
-            block.number,
-            endHash,
+        emit Registered(
+            _taskHash,
+            _solverPathRoot,
             executionDepth,
-            customEnvironmentHash,
-            msg.sender
+            _result
         );
-        bonds[msg.sender] += bondAmount;
-
-        emit Registered(executionId, msg.sender, codeContractAddress, _callData);
     }
 
-    /**
-      * @dev dispute is called by challenger to start a new dispute
-      *     assumed that challenger's execution tree is of the same depth as solver's
-      *     in case challenger's tree is shallower, he should use node with zero hash to make it deeper
-      *     in case challenger's tree is deeper, he should submit only the left subtree with the same depth with solver's
-      */
-    function dispute(address codeContractAddress, bytes memory _callData, bytes32 endHash)
+    function dispute(bytes32 _solverPathRoot, bytes32 _challengerPathRoot, EVMParameters memory _parameters)
```

Comment:
i'm missing here:
- `_taskHash` to be able to identify the task
- `bytes32[] memory _resultProof` a proof from `_challengerPathRoot` to `_result`
- `bytes memory _result` an alternative result

the result is mandatory, as it might neet to replace the challenged result [according to this flow](https://docs.google.com/drawings/d/1IWeGNEwm8EhQOxaM31IBFG7zSHDrfVHJdza_5Bpljcc).

i would like to get rid of:
- `EVMParameters memory _parameters` 

i know it is needed to initialize the dispute. can we not store the necessary data in the Task struct?

Reply:
> bytes32[] memory _resultProof a proof from _challengerPathRoot to _result

👍 

> i know it is needed to initialize the dispute. can we not store the necessary data in the Task struct?

We can store the EVMParameters, but I see no good reason to do that. We just massively increase the gas cost if we do that.